python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,810 @@
+                    f'You configured {gpu_per_task} GPU cores for each spark task, but in '
+                    f'XGBoost training, every Spark task will only use one GPU core.'
+                )
+
+
+class _XgboostEstimator(Estimator, _XgboostParams, MLReadable, MLWritable):
+    def __init__(self):
+        super().__init__()
+        self._set_xgb_params_default()
+        self._set_fit_params_default()
+        self._set_predict_params_default()
+        self._set_distributed_params()
+
+    def setParams(self, **kwargs):
+        _user_defined = {}
+        for k, v in kwargs.items():
+            if self.hasParam(k):
+                self._set(**{str(k): v})
+            else:
+                _user_defined[k] = v
+        _defined_args = self.getOrDefault(self.getParam("arbitraryParamsDict"))
+        _defined_args.update(_user_defined)
+        self._set(**{"arbitraryParamsDict": _defined_args})
+
+    @classmethod
+    def _pyspark_model_cls(cls):
+        """
+        Subclasses should override this method and
+        returns a _XgboostModel subclass
+        """
+        raise NotImplementedError()
+
+    def _create_pyspark_model(self, xgb_model):
+        return self._pyspark_model_cls()(xgb_model)
+
+    @classmethod
+    def _convert_to_classifier(cls, booster):
+        clf = XGBClassifier()
+        clf._Booster = booster
+        return clf
+
+    @classmethod
+    def _convert_to_regressor(cls, booster):
+        reg = XGBRegressor()
+        reg._Booster = booster
+        return reg
+
+    def _convert_to_model(self, booster):
+        if self._xgb_cls() == XGBRegressor:
+            return self._convert_to_regressor(booster)
+        elif self._xgb_cls() == XGBClassifier:
+            return self._convert_to_classifier(booster)
+        else:
+            return None  # check if this else statement is needed.
+
+    def _query_plan_contains_valid_repartition(self, query_plan,
+                                               num_partitions):
+        """
+        Returns true if the latest element in the logical plan is a valid repartition
+        """
+        start = query_plan.index("== Optimized Logical Plan ==")
+        start += len("== Optimized Logical Plan ==") + 1
+        num_workers = self.getOrDefault(self.num_workers)
+        if query_plan[start:start + len("Repartition")] == "Repartition" and \
+                num_workers == num_partitions:
+            return True
+        return False
+
+    def _repartition_needed(self, dataset):
+        """
+        We repartition the dataset if the number of workers is not equal to the number of
+        partitions. There is also a check to make sure there was "active partitioning"
+        where either Round Robin or Hash partitioning was actively used before this stage.
+        """
+        if self.getOrDefault(self.force_repartition):
+            return True
+        try:
+            num_partitions = dataset.rdd.getNumPartitions()
```

Comment:
I guess if not forcing repartition, there may be some data skew issue.

Reply:
yes that's the reason we do repartitioning.

Reply:
Yeah, force_repartition is defaulting to false, so if the users run their application and feel pretty slow, how they know the performance issue is caused by "data skew" or "xgboost native training itself"

Reply:
This is an issue, what about turn on repartition by default ?

Reply:
Yeah, I agree with you, let's default to true.

Reply:
@wbo4958 
On second thought, we have `_repartition_needed` , when `force_repartition` is False, it will check whether the dataframe need repartition, if it find the dataset has the same partition number with num_worker and has a "repartition" operator as the last step of query plan, then it will skip repartition, if check failed it will still do repartition.

Reply:
Looks good to me. Sorry for overloop